### PR TITLE
rfs: T4689: Support RFS(Receive Flow Steering)

### PR DESCRIPTION
**Pull request by jack9603301**
## Change Summary

Support RFS(Receive Flow Steering)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4689

## Component(s) name

rfs

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: smoketest/scripts/cli/test_interfaces_ethernet.py [hunk 1]
```diff
@@ -183,7 +184,27 @@ def test_offloading_rps(self):
...
+        for interface in self._interfaces:
+            self.cli_set(self._base_path + [interface, 'offload', 'rfs'])
+        self.cli_commit()
+
+        for interface in self._interfaces:
+            queues = glob(f'/sys/class/net/{interface}/queues/rx-*')
+            rfs_flow = 32768/queues
+            for i in range(0,queues):
+                flows = read_file(f'/sys/class/net/{interface}/queues/rx-{i}/rps_cpus')
+                if int(flows) == rfs_flow:
```
Comment (c-po) [inline smoketest/scripts/cli/test_interfaces_ethernet.py]: Please use self.assertEqual()

File: python/vyos/ifconfig/ethernet.py [hunk 2]
```diff
@@ -258,6 +259,21 @@ def set_rps(self, state):
...
 
+    def set_rps(self, state):
+        rfs_flow = 0
+        global_rfs_flow = 0
+        queues = glob(f'/sys/class/net/{ifname}/queues/rx-*')
+        if state:
+            global_rfs_flow = 32768
+            rfs_flow = 32768/queues
+
+        call('echo {global_rfs_flow} > /proc/sys/net/core/rps_sock_flow_entries')
```
Comment (c-po) [inline python/vyos/ifconfig/ethernet.py]: Missing f for format strings

File: python/vyos/ifconfig/ethernet.py [hunk 3]
```diff
@@ -258,6 +259,21 @@ def set_rps(self, state):
...
+
+        for i in range(0,queues):
+            call('echo {rfs_flow} > /sys/class/net/{ifname}/queues/rx-{i}/rps_flow_cnt')
```
Comment (c-po) [inline python/vyos/ifconfig/ethernet.py]: Missing f for format strings

File: python/vyos/ifconfig/ethernet.py [hunk 4]
```diff
@@ -258,6 +259,21 @@ def set_rps(self, state):
         # send bitmask representation as hex string without leading '0x'
         return self.set_interface('rps', rps_cpus)
 
+    def set_rps(self, state):
+        rfs_flow = 0
+        global_rfs_flow = 0
+        queues = glob(f'/sys/class/net/{ifname}/queues/rx-*')
+        if state:
+            global_rfs_flow = 32768
+            rfs_flow = 32768/queues
```
Comment (c-po) [inline python/vyos/ifconfig/ethernet.py]: Why not reference global_rfs_flows here?

**Review by c-po — CHANGES_REQUESTED** [review]

File: python/vyos/ifconfig/ethernet.py (hunk 4 above)
Comment (jack9603301) [reply python/vyos/ifconfig/ethernet.py]: done

File: python/vyos/ifconfig/ethernet.py (hunk 3 above)
Comment (jack9603301) [reply python/vyos/ifconfig/ethernet.py]: done

File: python/vyos/ifconfig/ethernet.py (hunk 2 above)
Comment (jack9603301) [reply python/vyos/ifconfig/ethernet.py]: done

File: smoketest/scripts/cli/test_interfaces_ethernet.py (hunk 1 above)
Comment (jack9603301) [reply smoketest/scripts/cli/test_interfaces_ethernet.py]: done

**Review by c-po — APPROVED** [review]

**Comment by c-po** [comment]
PR has been reverted due to multiple bugs in the implementation that caused images to fail to even boot up.

```python
Traceback (most recent call last):
  File "/usr/libexec/vyos/conf_mode/interfaces-ethernet.py", line 220, in <module>
    apply(c)
  File "/usr/libexec/vyos/conf_mode/interfaces-ethernet.py", line 209, in apply
    e.update(ethernet)
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/ethernet.py", line 362, in update
    self.set_rfs(dict_search('offload.rfs', config) != None)
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/ethernet.py", line 265, in set_rfs
    queues = glob(f'/sys/class/net/{ifname}/queues/rx-*')
NameError: name 'ifname' is not defined
```

Fixing this 
```diff
-        queues = glob(f'/sys/class/net/{ifname}/queues/rx-*')
+        queues = glob(f'/sys/class/net/{self.ifname}/queues/rx-*')
```

resulted in 

```python
Traceback (most recent call last):
  File "/usr/libexec/vyos/conf_mode/interfaces-ethernet.py", line 220, in <module>
    apply(c)
  File "/usr/libexec/vyos/conf_mode/interfaces-ethernet.py", line 209, in apply
    e.update(ethernet)
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/ethernet.py", line 362, in update
    self.set_rfs(dict_search('offload.rfs', config) != None)
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/ethernet.py", line 268, in set_rfs
    rfs_flow = global_rfs_flow/queues
TypeError: unsupported operand type(s) for /: 'int' and 'list'
```

Fixing this

```diff
-        queues = glob(f'/sys/class/net/{ifname}/queues/rx-*')
+        queues = len(glob(f'/sys/class/net/{self.ifname}/queues/rx-*'))
```

resulted in 

```python
Traceback (most recent call last):
  File "/usr/libexec/vyos/conf_mode/interfaces-ethernet.py", line 220, in <module>
    apply(c)
  File "/usr/libexec/vyos/conf_mode/interfaces-ethernet.py", line 209, in apply
    e.update(ethernet)
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/ethernet.py", line 362, in update
    self.set_rfs(dict_search('offload.rfs', config) != None)
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/ethernet.py", line 270, in set_rfs
    call(f'echo {global_rfs_flow} > /proc/sys/net/core/rps_sock_flow_entries')
NameError: name 'call' is not defined
```

Apparendtly this code was NEVER run through manual and also smoketesting 👎

**Comment by jack9603301** [comment]
Sorry, I will resubmit the PR after testing and improving

**Pull request merged**